internal/storage/index/builder.go

```diff
@@ -242,9 +254,24 @@ func (idx *indexBuilder) addPolicy(file string, srcCtx parser.SourceCtx, p polic
 	idx.fileToModID[file] = p.ID
 	idx.modIDToFile[p.ID] = file
 	delete(idx.missing, p.ID)
-	delete(idx.missingScopes, p.ID)
-
 	idx.stats.add(p)
+	if ids, ok := idx.deleteFromMissingScopes[p.ID]; ok {
```

Comment:
I am struggling to keep track of the juggling of entries between _four_ maps 😄 I think it can be simpler.


- Find the required ancestors for each policy passed to the `addPolicy` method
- If there are any ancestors that we haven't seen yet, add them to a map. The key is the module ID of the missing ancestor. The value is a set of FQNs that are descendants of it.
  Say you're processing `resource.foo.v1/a.b.c.d` and find that scopes `a.b` and `a.b.c` are missing. The map contents would be:
     ```go
      missingScopes["resource.foo.v1/a.b"] = map[string]struct{}{"resource.foo.v1/a.b.c.d": {}}
      missingScopes["resource.foo.v1/a.b.c"] = map[string]struct{}{"resource.foo.v1/a.b.c.d": {}}
     ```
- If there's an entry for the current policy in `missingScopes`, delete that because we know that the policy now exists
- At the end, create an error message for each item left in the `missingScopes` map. The message would be "scoped policy x is not found but is required by descendant policies a, b and c".